File: internal/audit/audit.go
package audit

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"sync"
	"time"
)

type Log struct {
	mu     sync.Mutex
	f      *os.File
	failed bool
}

type Event struct {
	Event              string `json:"event"`
	Timestamp          string `json:"timestamp"`
	CallID             string `json:"call_id,omitempty"`
	PublicTool         string `json:"public_tool,omitempty"`
	ServerID           string `json:"server_id,omitempty"`
	DownstreamToolName string `json:"downstream_tool_name,omitempty"`
	RegistryGeneration uint64 `json:"registry_generation,omitempty"`
	PolicyDigest       string `json:"policy_digest,omitempty"`
	Outcome            string `json:"outcome,omitempty"`
	DurationMS         int64  `json:"duration_ms,omitempty"`
}

func Open(path string) (*Log, error) {
	if info, e := os.Stat(path); e == nil && info.Mode().Perm()&0077 != 0 {
		return nil, fmt.Errorf("audit file permissions must not grant group or other access")
	} else if e != nil && !os.IsNotExist(e) {
		return nil, fmt.Errorf("stat audit: %w", e)
	}
	f, e := os.OpenFile(path, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0600)
	if e != nil {
		return nil, fmt.Errorf("open audit: %w", e)
	}
	l := &Log{f: f}
	if e = l.AppendReady(); e != nil {
		f.Close()
		return nil, e
	}
	return l, nil
}
func (l *Log) AppendReady() error {
	return l.Append(Event{Event: "audit_ready"})
}

func (l *Log) Append(event Event) error {
	l.mu.Lock()
	defer l.mu.Unlock()
	if l.failed || l.f == nil {
		return fmt.Errorf("audit unavailable")
	}
	if event.Timestamp == "" {
		event.Timestamp = time.Now().UTC().Format(time.RFC3339Nano)
	}
	b, err := json.Marshal(event)
	if err != nil {
		return fmt.Errorf("encode audit event: %w", err)
	}
	b = append(b, '\n')
	n, err := l.f.Write(b)
	if err == nil && n != len(b) {
		err = io.ErrShortWrite
	}
	if err == nil {
		err = l.f.Sync()
	}
	if err != nil {
		l.failed = true
		return fmt.Errorf("append audit event: %w", err)
	}
	return nil
}

func (l *Log) Available() bool {
	l.mu.Lock()
	defer l.mu.Unlock()
	return !l.failed && l.f != nil
}

func (l *Log) Close() error {
	l.mu.Lock()
	defer l.mu.Unlock()
	if l.f == nil {
		return nil
	}
	err := l.f.Close()
	l.f = nil
	l.failed = true
	return err
}

File: src/conduit/audit.py
"""Durable append-only audit log used as the authorization boundary."""

from __future__ import annotations

import asyncio
import json
import os
import stat
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

from .errors import AuditUnavailable


class AuditLog:
    """Serialize JSONL writes and fsync every accepted authorization event."""

    def __init__(self, path: str) -> None:
        target = Path(path)
        try:
            if target.exists() and stat.S_IMODE(target.stat().st_mode) & 0o077:
                raise AuditUnavailable("audit file permissions must not grant group or other access")
            self._file = os.fdopen(os.open(target, os.O_WRONLY | os.O_APPEND | os.O_CREAT, 0o600), "a", encoding="utf-8")
        except OSError as exc:
            raise AuditUnavailable(f"open audit: {exc}") from exc
        self._lock = asyncio.Lock()
        self._failed = False
        self._closed = False

    async def ready(self) -> None:
        await self.append({"event": "audit_ready"})

    async def append(self, event: dict[str, Any]) -> None:
        async with self._lock:
            if self._failed or self._closed:
                raise AuditUnavailable("audit unavailable")
            payload = dict(event)
            payload.setdefault("timestamp", datetime.now(UTC).isoformat().replace("+00:00", "Z"))
            try:
                self._file.write(json.dumps(payload, separators=(",", ":"), ensure_ascii=False) + "\n")
                self._file.flush()
                os.fsync(self._file.fileno())
            except (OSError, TypeError, ValueError) as exc:
                self._failed = True
                raise AuditUnavailable("audit unavailable") from exc

    @property
    def available(self) -> bool:
        return not self._failed and not self._closed

    async def close(self) -> None:
        async with self._lock:
            if not self._closed:
                self._file.close()
                self._closed = True
                self._failed = True
